Added required __future__ imports

Sentient07 opened the pull request:
fix #3900

lamblin commented:
I have not looked in details yet. It is a good start, but:
- we should incorporate the script checking for the presence of imports so that it gets called in `theano/tests/test_flake8.py`. That may be done via an install step in Travis, I'm not sure exactly
- the imports should not be added a second time for files that already import them
- the list of files where the "division" import was not added would be actually very useful, so that we can investigate where the issues come from.

Sentient07 commented:
@lamblin , 

- yeah. Installing that package which you had suggested in the issue comment would do, but i doubt that would help. That script checks for all the missing ```__future__``` imports. Importing them wouldn't be an issue, but we need to cherry pick and remove the ones that raises error.

- The imports aren't added the second time. This is ensured if travis passes because, imports from ```__future__``` modules have to be in the first line of a file

- Yeah, i'll get the list by greping. It's mainly the issue raised by ```/``` operator, which performs classic division if ```division``` is imported from ```__future__```. Else, it performs floor division. I don't think it's a bad idea to replace with the ```//``` operator throughout for floor division for consistency purpose. Shall i open a new issue for that or work on this ?

Sentient07 commented:
@lamblin As far as i can see, these are the list of files that raises error when division is imported. 

- tensor/xlogx.py
- compile/tests/test_function_name.py (raises error with all the three imports)
- scalar/tests/test_basic.py
- scan_module/tests/test_scan.py

I'll branch out from this branch, replace the operator(with the imports intact) and check if the error occurs.

lamblin commented:
- I need we need an automatic way of detecting that for future files. We may need a white list for the moment, but I think the files where the future imports cause an error should be fixed, as it probably indicates that there is already a problem with python 3.

- Some files are cluttered with duplicate imports. For instance if you look at https://github.com/Theano/Theano/pull/3983/files#diff-2db7fbaa2b361e6cceb2861a00a98a16, the file contains:
```python
from __future__ import absolute_import, print_function, division
#!/usr/bin/env python
from __future__ import print_function
from __future__ import absolute_import
```
because `print_function` and `absolute_import` were already imported.
Also, you need to keep `#!/usr/bin/env python` as the first line so that it is correctly recognized as an executable by the OS. Python sees it as a comment, so it is fine to add the future imports afterwards.

- I think this is part of this ticket.

Sentient07 commented:
@lamblin should the files that cause the error be fixed in this PR, or should i open a new one ? I'll look into the duplication and would fix it in all the files.

lamblin commented:
I think they should either be fixed in this PR, or be fixed in a separate PR that should be merged _before_ this one.
I would favor the second option if you are comfortable with the rebasing / cherry-picking / conflicts that may result. Otherwise, it would be fine in this PR.

Sentient07 commented:
@lamblin yeah, i'll branch out from master, then make those changes and open a new PR ? Also, i need to keep rebasing this PR with almost every PR that gets merged with master I guess..

Sentient07 commented:
@lamblin It seems like ```/``` operation when performed after importing division from __future__ doesn't work on the type ```ScalarVariable```. I get a type error with that. But surprisingly, that's not the case with python 3. I've opened a new PR, https://github.com/Theano/Theano/pull/3999

nouiz commented:
Try this diff to fix the problem with ScalarVariable.

```
diff --git a/theano/scalar/basic.py b/theano/scalar/basic.py
index ab86c58..75a1b78 100644
--- a/theano/scalar/basic.py
+++ b/theano/scalar/basic.py
@@ -607,9 +607,9 @@ class _scalar_py_operators:
     if PY3:
         def __truediv__(self, other):
             return div_proxy(self, other)
-    else:
-        def __div__(self, other):
-            return div_proxy(self, other)
+
+    def __div__(self, other):
+        return div_proxy(self, other)

     def __floordiv__(self, other):
         return int_div(self, other)

```

On Fri, Feb 5, 2016 at 10:40 AM, Ramana Subramanyam <
notifications@github.com> wrote:

> @lamblin <https://github.com/lamblin> It seems like / operation when
> performed after importing division from *future* doesn't work on the type
> ScalarVariable. I get a type error with that. But surprisingly, that's
> not the case with python 3. I've opened a new PR, #3999
> <https://github.com/Theano/Theano/pull/3999>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3983#issuecomment-180406869>.
>

Sentient07 commented:
@lamblin yeah, i navigated to the file and fixed it. The fix was to make the ```__truediv__``` not only python 3 specific. I had updated it in the new PR.

Sentient07 commented:
@lamblin fixed the merge conflicts in this branch. Is this also ready for merge or do i need to check something else as well ?

lamblin commented:
So I think the remaining steps are:
- fixing the pep8 errors that probably got introduced during the conflict resolution
- Add the setup for flake8-future-import in travis.yml so that there is an automatic check

Sentient07 commented:
@lamblin , if i add that, it raises a warning/prints a message for the imports that are preset as well. Would that be an issue ?

lamblin commented:
@Sentient07 You can set up the checks to disable the warnings when imports are present. We already disable some error codes [here](https://github.com/Sentient07/Theano/blob/fdb5f565440922c9e61b360015950e984712e646/theano/tests/test_flake8.py#L25), so you could do it at the same place.

Sentient07 commented:
@lamblin, yeah sure. Will do that. Also, i guess travis isn't fixed. I'll check that, this and fix it before sleep :+1:

Sentient07 commented:
@lamblin this expects us to add the imports even to the empty ```__init__.py``` files as well. Shall I add or exclude those files ?

lamblin commented:
I think you can have a whitelist for empty files, and check instead that they stay empty.

Sentient07 commented:
@lamblin The travis is still failing. I went through the error message but i'm not understanding what's going wrong. I don't think it is from the changes that i have made. I ran the tests locally and it passed. Could you please help me with this ?

lamblin commented:
The errors seem unrelated. #4176 should fix them, I'll restart the build when things are fixed.

Sentient07 commented:
@lamblin Shall I force push for travis to start building again ?

lamblin commented:
The close/reopen should have done so, but someone stopped it.
I'll restart it when I can, please do not push.

Sentient07 commented:
@lamblin Sure.

abergeron commented:
Breakage fixed, new run in progress.

Sentient07 commented:
@abergeron , there is still one test failing in python 3.3, please check at your convenience. :)

lamblin commented:
Looks like a stochastic failure, I relaunched it and we will probably up the tolerance.

Sentient07 commented:
@lamblin looks like it needs a rebase. I'd do that and update the PR ?

lamblin commented:
I'll try to review today, and then you can do all changes + the rebase at once. That can ease the load of Travis.
Thanks,

lamblin commented:
I'll comment here because GitHub's diff view is overwhelmed and I can't do line comments:
- in `theano/tests/test_flake8.py`: can you add a comment explaining what the additional error codes mean? There was a comment for the existing ones.
- in `setup.cfg`, can you add the error codes there as well, in the `flake8` section?
- in `theano/sparse/tests/test_utils.py`, I think the reordering of imports is wrong: now you raise `SkipTest` before importing it, and you import things from `theano.sparse` even if `theano.sparse.enable_sparse` is False. If you need to make `flake8` happy, I would suggest adding `# noqa` at the end of the `if not ...` line.

Finally, I'm not sure if we should wait for after the 0.8 release to merge that. @nouiz? @abergeron?

nouiz commented:
As this don't add new feature or benefit to the release and it is very
big(so easy to miss a small change during review) I would prefer to wait
after the 0.8 release.

On Tue, Mar 8, 2016 at 7:42 PM, Pascal Lamblin <notifications@github.com>
wrote:

> I'll comment here because GitHub's diff view is overwhelmed and I can't do
> line comments:
>
>    - in theano/tests/test_flake8.py: can you add a comment explaining
>    what the additional error codes mean? There was a comment for the existing
>    ones.
>    - in setup.cfg, can you add the error codes there as well, in the
>    flake8 section?
>    - in theano/sparse/tests/test_utils.py, I think the reordering of
>    imports is wrong: now you raise SkipTest before importing it, and you
>    import things from theano.sparse even if theano.sparse.enable_sparse
>    is False. If you need to make flake8 happy, I would suggest adding #
>    noqa at the end of the if not ... line.
>
> Finally, I'm not sure if we should wait for after the 0.8 release to merge
> that. @nouiz <https://github.com/nouiz>? @abergeron
> <https://github.com/abergeron>?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Theano/Theano/pull/3983#issuecomment-194039985>.
>

Sentient07 commented:
@lamblin @nouiz  I've addressed the all comments to the best of my knowledge. I will keep rebasing this branch frequently, fixing the merge conflicts. Have a look at your convenience. Thanks.

lamblin commented:
Thanks, we will merge after the release, if you address the last remaining comment.
Thanks for your patience.

Sentient07 commented:
@lamblin I have addressed that PR and rebased.

nouiz commented:
Travis is failing. There is many errors. At least:

```
/home/travis/build/Theano/Theano/theano/tests/record.py:111:18: F821 undefined name 'Mode'

/home/travis/build/Theano/Theano/theano/tests/test_2nd_order_grads.py:5:1: F401 'unittest' imported but unused

/home/travis/build/Theano/Theano/theano/tests/test_record.py:2:1: F403 'from theano.tests.record import *' used; unable to detect undefined names


======================================================================

ERROR: Failure: NameError (name 'Mode' is not defined)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/loader.py", line 413, in loadTestsFromName

    addr.filename, addr.module)

  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/importer.py", line 47, in importFromPath

    return self.importFromDir(dir_path, fqname)

  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/importer.py", line 94, in importFromDir

    mod = load_module(part_fqname, fh, filename, desc)

  File "/home/travis/build/Theano/Theano/theano/tests/test_determinism.py", line 2, in <module>

    from theano.tests.record import RecordMode, Record

  File "/home/travis/build/Theano/Theano/theano/tests/record.py", line 111, in <module>

    class RecordMode(Mode):

NameError: name 'Mode' is not defined

======================================================================

ERROR: Failure: NameError (name 'Mode' is not defined)

----------------------------------------------------------------------

Traceback (most recent call last):

  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/loader.py", line 413, in loadTestsFromName

    addr.filename, addr.module)

  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/importer.py", line 47, in importFromPath

    return self.importFromDir(dir_path, fqname)

  File "/home/travis/miniconda2/envs/pyenv/lib/python2.6/site-packages/nose/importer.py", line 94, in importFromDir

    mod = load_module(part_fqname, fh, filename, desc)

  File "/home/travis/build/Theano/Theano/theano/tests/test_record.py", line 2, in <module>

    from theano.tests.record import *

  File "/home/travis/build/Theano/Theano/theano/tests/record.py", line 111, in <module>

    class RecordMode(Mode):

NameError: name 'Mode' is not defined

----------------------------------------------------------------------

Ran 998 tests in 1818.400s

FAILED (SKIP=98, errors=2)

```

As now the release is done, we try to merge the PR that was waiting.

thanks

Sentient07 commented:
@nouiz Rebased this branch last night. It was fine before that. some error has occurred while copy pasting during rebase.Sorry was a little busy and didn't give much attention to this PR.  Will figure out and fix this by tonight. Thanks.

Sentient07 commented:
@nouiz Three errors in flake8 test pertaining to imports, which as i guessed earlier was due to rebasing. I ran the failing tests locally and they pass.

lamblin commented on theano/tests/test_flake8.py:
```diff
@@ -30,8 +30,21 @@
+    "compile/tests/test_ops.py",
+    "compile/tests/test_pfunc.py",
+    "compile/tests/test_debugmode.py",
+    "compile/tests/test_profiling.py",
```
It is OK to add empty `__init__.py` file in the whitelist, but here you are adding more.
Maybe these files got removed from the whitelist in the mean time, which is why you still had them when rebasing.
In any case, you should remove these `compile/tests/test_*.py` from the whitelist.

lamblin commented on theano/tests/test_flake8.py:
```diff
@@ -126,15 +147,26 @@
+    "misc/tests/test_gnumpy_utils.py",
+    "misc/tests/test_pycuda_utils.py",
+    "misc/tests/test_cudamat_utils.py",
+    "misc/tests/test_pycuda_example.py",
```
Same here.

lamblin commented:
GitHub shows the comments as outdated, and I can't comment directly on the diff, but the following files should not be added back to the whitelist in `theano/tests/test_flake8.py`:

- "compile/tests/test_builders.py",
- "compile/tests/test_misc.py",
- "compile/tests/test_monitormode.py",
- "compile/tests/test_function_module.py",
- "compile/tests/test_shared.py",
- "compile/tests/test_ops.py",
- "compile/tests/test_pfunc.py",
- "compile/tests/test_debugmode.py",
- "compile/tests/test_profiling.py",
- "misc/tests/test_may_share_memory.py",
- "misc/tests/test_pycuda_theano_simple.py",
- "misc/tests/test_gnumpy_utils.py",
- "misc/tests/test_pycuda_utils.py",
- "misc/tests/test_cudamat_utils.py",
- "misc/tests/test_pycuda_example.py"

Sentient07 commented:
@lamblin this could be merged now ?

lamblin commented:
Looks good, thanks!
